java-checks/src/main/java/org/sonar/java/checks/regex/CanonEqFlagInRegexCheck.java

```diff
@@ -0,0 +1,136 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks.regex;
+
+import java.text.Normalizer;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.regex.Pattern;
+import org.sonar.check.Rule;
+import org.sonar.java.regex.RegexParseResult;
+import org.sonar.java.regex.ast.CharacterClassTree;
+import org.sonar.java.regex.ast.JavaCharacter;
+import org.sonar.java.regex.ast.PlainCharacterTree;
+import org.sonar.java.regex.ast.RegexBaseVisitor;
+import org.sonar.java.regex.ast.RegexTree;
+import org.sonar.java.regex.ast.SequenceTree;
+import org.sonar.plugins.java.api.semantic.MethodMatchers;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+
+@Rule(key = "S5854")
+public class CanonEqFlagInRegexCheck extends AbstractRegexCheck {
+
+  protected static final MethodMatchers STRING_MATCHES = MethodMatchers.create()
+    .ofTypes(JAVA_LANG_STRING)
+    .names("matches")
+    .addParametersMatcher(JAVA_LANG_STRING)
+    .build();
+
+  protected static final MethodMatchers STRING_REPLACE_ALL = MethodMatchers.create()
+    .ofTypes(JAVA_LANG_STRING)
+    .names("replaceAll")
+    .addParametersMatcher(JAVA_LANG_STRING, JAVA_LANG_STRING)
+    .build();
+
+  protected static final MethodMatchers STRING_REPLACE_FIRST = MethodMatchers.create()
+    .ofTypes(JAVA_LANG_STRING)
+    .names("replaceFirst")
+    .addParametersMatcher(JAVA_LANG_STRING, JAVA_LANG_STRING)
+    .build();
+
+  protected static final MethodMatchers PATTERN_MATCHES = MethodMatchers.create()
+    .ofTypes("java.util.regex.Pattern")
+    .names("matches")
+    .addParametersMatcher(JAVA_LANG_STRING, "java.lang.CharSequence")
+    .build();
+
+  @Override
+  public void checkRegex(RegexParseResult regexForLiterals, MethodInvocationTree mit) {
+    if (regexForLiterals.getInitialFlags().contains(Pattern.CANON_EQ)) {
+      return;
+    }
+    PlainCharacterVisitor visitor = new PlainCharacterVisitor();
+    visitor.visit(regexForLiterals);
+
+    if (!visitor.subjectToNormalization.isEmpty()) {
+      String endOfMessage;
+      if (STRING_MATCHES.matches(mit) || PATTERN_MATCHES.matches(mit)) {
+        endOfMessage = "\"Pattern.compile(regex, CANON_EQ).matcher(input).matches()\"";
+      } else if (STRING_REPLACE_ALL.matches(mit)) {
+        endOfMessage = "\"Pattern.compile(pattern, CANON_EQ).matcher(input).replaceAll(replacement)\"";
+      } else if (STRING_REPLACE_FIRST.matches(mit)) {
+        endOfMessage = "\"Pattern.compile(pattern, CANON_EQ).matcher(input).replaceFirst(replacement)\"";
+      } else {
+        endOfMessage = "this pattern";
+      }
```

Comment:
Nice catch that enabling the flag might require using a different method and to include this in the message. :+1: